### PR TITLE
feat: rewrite e2e test(rewrite-e2e-test-route-service-upstream-test) with ginkgo

### PR DESCRIPTION
**Pull request by Jaycean**
Please answer these questions before submitting a pull request, **or your PR will get closed**.

**Why submit this pull request?**

- [ ] Bugfix
- [x] New feature provided
- [ ] Improve performance
- [ ] Backport patches

**What changes will this PR take into?**

feat: rewrite e2e test(rewrite-e2e-test-route-service-upstream-test) with ginkgo

**Related issues**

#1500

**Checklist:**

- [x] Did you explain what problem does this PR solve? Or what new features have been added?
- [ ] Have you added corresponding test cases?
- [ ] Have you modified the corresponding document?
- [x] Is this PR backward compatible? If it is not backward compatible, please discuss on the mailing list first

### CODE REVIEW

File: api/test/e2enew/route/route_service_upstream_test.go [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+
+	"github.com/onsi/ginkgo"
+	"github.com/onsi/ginkgo/extensions/table"
+	"github.com/onsi/gomega"
+
+	"github.com/apisix/manager-api/test/e2enew/base"
+)
+
+var _ = ginkgo.Describe("route invalid service and service", func() {
+	table.DescribeTable("test route invalid service and service",
```
Comment (tokers) [inline api/test/e2enew/route/route_service_upstream_test.go:31]: The title is obsecure.

File: api/test/e2enew/route/route_service_upstream_test.go [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+			Path:   "/apisix/admin/routes/r1",
+			Body: `{
+				"name": "route1",
+				"uri": "/hello_",
+				"service_id": "not-exists"
+			}`,
+			Headers:      map[string]string{"Authorization": base.GetToken()},
+			ExpectStatus: http.StatusBadRequest,
+		}),
+		table.Entry("hit invalid route on data plane", base.HttpTestCase{
```
Comment (tokers) [inline api/test/e2enew/route/route_service_upstream_test.go:47]: Bad title, it not an invalid route, actually it doesn't exist.

File: api/test/e2enew/route/route_service_upstream_test.go [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+			Path:   "/apisix/admin/routes/r1",
+			Body: `{
+				"name": "route1",
+				"uri": "/hello_",
+				"upstream_id": "not-exists"
+			}`,
+			Headers:      map[string]string{"Authorization": base.GetToken()},
+			ExpectStatus: http.StatusBadRequest,
+		}),
+		table.Entry("hit invalid route on data plane", base.HttpTestCase{
```
Comment (tokers) [inline api/test/e2enew/route/route_service_upstream_test.go:66]: Ditto.

File: api/test/e2enew/route/route_service_upstream_test.go [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+			Body: `{
+				"name": "route1",
+				"uri": "/hello_",
+				"service_id": "not-exists-service",
+				"upstream_id": "not-exists-upstream"
+			}`,
+			Headers:      map[string]string{"Authorization": base.GetToken()},
+			ExpectStatus: http.StatusBadRequest,
+		}),
+		table.Entry("hit invalid route on data plane", base.HttpTestCase{
```
Comment (tokers) [inline api/test/e2enew/route/route_service_upstream_test.go:86]: Ditto.

File: api/test/e2enew/route/route_service_upstream_test.go [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+				"service_id": "not-exists"
+			}`,
+			Headers:      map[string]string{"Authorization": base.GetToken()},
+			ExpectStatus: http.StatusBadRequest,
+		}),
+		table.Entry("hit invalid route on data plane", base.HttpTestCase{
+			Object:       base.APISIXExpect(),
+			Method:       http.MethodGet,
+			Path:         "/hello_",
+			ExpectStatus: http.StatusNotFound,
```
Comment (tokers) [inline api/test/e2enew/route/route_service_upstream_test.go:51]: The response body should also be checked as we must ensure the `404` is returned by APISIX and caused by "route not found".

File: api/test/e2enew/route/route_service_upstream_test.go [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+			ExpectStatus: http.StatusOK,
+		}),
+		table.Entry("remove service", base.HttpTestCase{
+			Object:       base.ManagerApiExpect(),
+			Method:       http.MethodDelete,
+			Path:         "/apisix/admin/services/200",
+			Headers:      map[string]string{"Authorization": base.GetToken()},
+			ExpectStatus: http.StatusOK,
+		}),
+		table.Entry("hit deleted route", base.HttpTestCase{
```
Comment (tokers) [inline api/test/e2enew/route/route_service_upstream_test.go:180]: Bad title, how can we hit a deleted route?

File: api/test/e2enew/route/route_service_upstream_test.go [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+			ExpectStatus: http.StatusOK,
+		}),
+		table.Entry("make sure the route has not created", base.HttpTestCase{
+			Object:       base.APISIXExpect(),
+			Method:       http.MethodGet,
+			Path:         "/server_port",
+			ExpectStatus: http.StatusNotFound,
+			Sleep:        base.SleepTime,
+		}),
+		table.Entry("create route using the service just created", base.HttpTestCase{
```
Comment (tokers) [inline api/test/e2enew/route/route_service_upstream_test.go:228]: Bad title. Should be "create route using the upstream just created".

File: api/test/e2enew/route/route_service_upstream_test.go [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+			ExpectStatus: http.StatusOK,
+		}),
+		table.Entry("remove upstream", base.HttpTestCase{
+			Object:       base.ManagerApiExpect(),
+			Method:       http.MethodDelete,
+			Path:         "/apisix/admin/upstreams/1",
+			Headers:      map[string]string{"Authorization": base.GetToken()},
+			ExpectStatus: http.StatusOK,
+		}),
+		table.Entry("hit deleted route", base.HttpTestCase{
```
Comment (tokers) [inline api/test/e2enew/route/route_service_upstream_test.go:270]: Bad title, how can we hit a deleted route?

**Review by tokers — CHANGES_REQUESTED** [review]

File: api/test/e2enew/route/route_service_upstream_test.go [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+			ExpectStatus: http.StatusBadRequest,
+		}),
+		table.Entry("hit invalid route on data plane", base.HttpTestCase{
+			Object:       base.APISIXExpect(),
+			Method:       http.MethodGet,
+			Path:         "/hello_",
+			ExpectStatus: http.StatusNotFound,
+			Sleep:        base.SleepTime,
+		}),
+	)
```
Comment (starsz) [inline api/test/e2enew/route/route_service_upstream_test.go:93]: Maybe we can test this situation:

`service that contains not-exists-upstream`

File: api/test/e2enew/route/route_service_upstream_test.go [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+			ExpectStatus: http.StatusOK,
+		}),
+		table.Entry("hit deleted route", base.HttpTestCase{
+			Object:       base.APISIXExpect(),
+			Method:       http.MethodGet,
+			Path:         "/server_port",
+			ExpectStatus: http.StatusNotFound,
+		}),
+	)
+})
```
Comment (starsz) [inline api/test/e2enew/route/route_service_upstream_test.go:277]: Maybe we can test this situation:

`service that contains upstream`

File: api/test/e2enew/route/route_service_upstream_test.go (hunk 1 above)
Comment (Jaycean) [reply api/test/e2enew/route/route_service_upstream_test.go]: done.

File: api/test/e2enew/route/route_service_upstream_test.go (hunk 2 above)
Comment (Jaycean) [reply api/test/e2enew/route/route_service_upstream_test.go]: done.

File: api/test/e2enew/route/route_service_upstream_test.go (hunk 3 above)
Comment (Jaycean) [reply api/test/e2enew/route/route_service_upstream_test.go]: done.

File: api/test/e2enew/route/route_service_upstream_test.go (hunk 4 above)
Comment (Jaycean) [reply api/test/e2enew/route/route_service_upstream_test.go]: done.

File: api/test/e2enew/route/route_service_upstream_test.go (hunk 5 above)
Comment (Jaycean) [reply api/test/e2enew/route/route_service_upstream_test.go:59]: done.

File: api/test/e2enew/route/route_service_upstream_test.go (hunk 6 above)
Comment (Jaycean) [reply api/test/e2enew/route/route_service_upstream_test.go]: done.

File: api/test/e2enew/route/route_service_upstream_test.go (hunk 7 above)
Comment (Jaycean) [reply api/test/e2enew/route/route_service_upstream_test.go]: done.

File: api/test/e2enew/route/route_service_upstream_test.go (hunk 8 above)
Comment (Jaycean) [reply api/test/e2enew/route/route_service_upstream_test.go]: done.

File: api/test/e2enew/route/route_service_upstream_test.go (hunk 9 above)
Comment (Jaycean) [reply api/test/e2enew/route/route_service_upstream_test.go:134]: done.

File: api/test/e2enew/route/route_service_upstream_test.go (hunk 10 above)
Comment (Jaycean) [reply api/test/e2enew/route/route_service_upstream_test.go:425]: done.

**Comment by Jaycean** [comment]
cc @juzhiyuan @liuxiran FE CI Failed. PTAL

**Review by nic-chen — APPROVED** [review]

**Review by tokers — APPROVED** [review]

**Review by gxthrj — APPROVED** [review]

**Pull request merged**